lightning-invoice/src/payment.rs

```diff
@@ -1397,18 +1564,263 @@ mod tests {
+				Some(vec![
+					Err(ChannelUnavailable { err: "abc".to_string() }), Err(MonitorUpdateFailed)
+				]))
+			.expect_send(Amount::ForInvoice(final_value_msat));
+
+		let router = TestRouter {};
+		let scorer = RefCell::new(TestScorer::new());
+		let logger = TestLogger::new();
+		let invoice_payer =
+			InvoicePayer::new(&payer, router, &scorer, &logger, event_handler, Retry::Attempts(0));
+
+		invoice_payer.pay_invoice(&invoice_to_pay).unwrap();
+		let inflight_map = invoice_payer.create_inflight_map();
+
+		// Only the second path, which failed with `MonitorUpdateFailed` should be added to our
+		// inflight map because retries are disabled.
+		assert_eq!(inflight_map.len(), 1);
+	}
+
+	#[test]
+	fn accounts_for_some_inflight_htlcs_sent_during_partial_failure() {
+		let event_handled = core::cell::RefCell::new(false);
+		let event_handler = |_: &_| { *event_handled.borrow_mut() = true; };
+
+		let payment_preimage = PaymentPreimage([1; 32]);
+		let invoice_to_pay = invoice(payment_preimage);
+		let final_value_msat = invoice_to_pay.amount_milli_satoshis().unwrap();
+
+		let retry = TestRouter::retry_for_invoice(&invoice_to_pay);
+		let payer = TestPayer::new()
+			.fails_with_partial_failure(
+				retry.clone(), OnAttempt(1),
+				Some(vec![
+					Ok(()), Err(MonitorUpdateFailed)
+				]))
+			.expect_send(Amount::ForInvoice(final_value_msat));
+
+		let router = TestRouter {};
+		let scorer = RefCell::new(TestScorer::new());
+		let logger = TestLogger::new();
+		let invoice_payer =
+			InvoicePayer::new(&payer, router, &scorer, &logger, event_handler, Retry::Attempts(0));
+
+		invoice_payer.pay_invoice(&invoice_to_pay).unwrap();
+		let inflight_map = invoice_payer.create_inflight_map();
+
+		// All paths successful, hence we check of the existence of all 4 hops.
+		assert_eq!(inflight_map.len(), 4);
+	}
+
 	struct TestRouter;
 
 	impl TestRouter {
 		fn route_for_value(final_value_msat: u64) -> Route {
 			Route {
 				paths: vec![
-					vec![RouteHop {
-						pubkey: PublicKey::from_slice(&hex::decode("02eec7245d6b7d2ccb30380bfbe2a3648cd7a942653f5aa340edcea1f283686619").unwrap()[..]).unwrap(),
-						channel_features: ChannelFeatures::empty(),
-						node_features: NodeFeatures::empty(),
-						short_channel_id: 0, fee_msat: final_value_msat / 2, cltv_expiry_delta: 144
-					}],
+					vec![
+						RouteHop {
+							pubkey: PublicKey::from_slice(&hex::decode("02eec7245d6b7d2ccb30380bfbe2a3648cd7a942653f5aa340edcea1f283686619").unwrap()[..]).unwrap(),
+							channel_features: ChannelFeatures::empty(),
+							node_features: NodeFeatures::empty(),
+							short_channel_id: 0,
+							fee_msat: 10,
+							cltv_expiry_delta: 0
+						},
+						RouteHop {
+							pubkey: PublicKey::from_slice(&hex::decode("0324653eac434488002cc06bbfb7f10fe18991e35f9fe4302dbea6d2353dc0ab1c").unwrap()[..]).unwrap(),
+							channel_features: ChannelFeatures::empty(),
+							node_features: NodeFeatures::empty(),
+							short_channel_id: 1,
```

Comment:
So, technically, this isn't possible if the second path has a single hop from the payer to the recipient with the same `short_channel_id`, as it would mean the paths have different recipients. Then again, that was the case prior to your changes, so maybe shouldn't worry about making this realistic. It just confused me as to why the last test above has 4 hops when there should have been an overlap. But now it makes sense.

Reply:
hm, why will they have different recipients if the pubkey is the same?

Reply:
The last hop in each path is the recipient, which are different here. Additionally, `TestPayer::node_id` returns pubkey `"032..."`, which is same as the node through hop with `short_channel_id` 1. This means that this hop connects different nodes on each path, which wouldn't be possible in practice.

So in actuality:
- the first path contains a cycle back to the sender on the second hop
- the second path has one hop going directly back to the sender (itself), but with a `short_channel_id` that conflicts with a hop in the first path 
- we see four hops in the map instead of three because on the first path the hop's key is `(1, true)` (i.e., `"02e..." < "032..."`) but on the second path it is `(1, false)` (i.e., `"032..." < "032..."`)

FWIW, this partly predates your change. If we want this to be both realistic and have appropriate test coverage, we should:
- Change `TestPayer::node_id` return a pubkey not used in any of the hops.
- Change the second path to have three hops, at least one of which doesn't overlap with the first path. Maybe two hops would suffice depending on how the tests are formulated.
- Re-number the hop's `short_channel_id` starting from 1 since your code uses 0 for a fake hop, which currently conflicts with the first path's first hop, though doesn't cause any problems.

But this change may affect preexisting tests.

Reply:
ahhh, gotcha. I didn't notice that about `TestPayer::node_id`. Makes sense now. I'll make the changes.

Reply:
hm, I just checked, it looks like `TestPayer::node_id` doesn't actually return `"32..."`?

payer pubkey:
<img width="390" alt="image" src="https://user-images.githubusercontent.com/5944973/187055809-a3e2bb02-d81a-49d7-a8dc-04415202dbdc.png">

1st path, 2nd hop pubkey
<img width="494" alt="image" src="https://user-images.githubusercontent.com/5944973/187055851-ffcb72f0-2507-4fea-a22b-315fb69ec157.png">


`TestPayer::node_id` call:

```
		fn node_id(&self) -> PublicKey {
			let secp_ctx = Secp256k1::new();
			PublicKey::from_secret_key(&secp_ctx, &SecretKey::from_slice(&[42; 32]).unwrap())
		}
```

am i missing something?

Reply:
> since your code uses 0 for a fake hop

we actually never hardcode any of the channel ids. if you're referring to the `0` in `create_inflight_map`, that's meant to represent a dummy hop to our source, so that we can get the payer's node information when iterating backward of the path to construct the map

Reply:
think here's what I'm going to do:

```
       —>—————>— [ 029e... ] ————>—————
      /                                                                           \
   (3)                                                                         (4)
    /                                                                                \
[ Us ] --(0)--> [ 02ee... ] --(1)--> [ 032... ] --2--> [ 027f... ]
```

1. Adjust the last path to have 2 hops, one to `029e...` and then `027f...`
2. Make sure all the scids are unique
3. Update tests

Reply:
Resolved in https://github.com/lightningdevkit/rust-lightning/pull/1643/commits/1abd3df8ab460d4b7fffd32dae5ae7419a8739d5

Reply:
> hm, I just checked, it looks like `TestPayer::node_id` doesn't actually return `"32..."`?
> 

Whoops, I was using `42` hex instead of `42` decimal. 😅 But note that the first byte as `44` decimal corresponds to the last byte as `2c` hexadecimal. The full pubkey should be:

`035be5e9478209674a96e60f1f037f6176540fd001fa1d64694770c56a7709c42c`